### ISSUE TITLE
single precision not supported with MKL on macOS Sierra?

### ISSUE THREAD

**Issue by ysmiraak** [uncomplicate/neanderthal#30]
i really love this library, and i'm migrating from tensorflow. because i'm used to f32, i'm walking through the tutorials with single precision data, and encountered some unexpected behaviors:

```
(dot (dv (range 3)) (dv (range 3))) => 5.0
(dot (fv (range 3)) (fv (range 3))) => 0.0
```

```
(asum (dv (range 3))) => 3.0
(asum (fv (range 3))) => 0.0
```

```
(nrm1 (dv (range 3))) => 3.0
(nrm1 (fv (range 3))) => 0.0

(nrm2 (dv (range 3))) => 2.23606797749979
(nrm2 (fv (range 3))) => -1.0842022E-19
```

namely `dot`, `asum`, `nrm1`, `nrm2` do not work for single precision.
(`mv`, `mm`, `sum`, `nrmi` work properly.)

i wonder, is this a bug, or intentional in that single precision CPU computing is not supported?

i'm on macOS Sierra `10.12.6`. i installed MKL as instructed, and moved all the MKL dylibs & libiomp5 to `/usr/local/lib/`. i'm using neanderthal `0.14.0` and clojure `1.9.0-alpha17`.

**Comment by blueberry** [comment]
I've just tried this in Linux and Mac OS X El Capitan (10.11). In Linux, it works as expected, while in ElCapitan I can reproduce your results.

Since it works as expected in Linux, and the Neanderthal code is the same for Linux and Mac OS X, i assume that the issue is either in MKL on OSX itself, or the GCC compiler on OSX, or some combination of those. I'll try to find out what's the issuse, but until then, I can only advise you to use double precision on OSX, or check the particular single precision functions that you want to use (and if they do not work properly, use double precision).

**Comment by blueberry** [comment]
It would be helpful in identifying the issue if more OSX users could try this with their different versions of OSX and MKL and post here the versions they have and the results they get. I assume you use the last version of MKL?

**Comment by ysmiraak** [comment]
yes, i installed the latest version of MKL `2017 update 3`.

**Comment by blueberry** [comment]
These kinds of issues (https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/291179) could also be due to Apple's linker, which may or may not work properly depending on GCC version. If any of the OSX users have time to tinker with some knobs during neanderthal-mkl compilation, that might lead to solution.

**Comment by ysmiraak** [comment]
sorry i won't be of any help here. i have little knowledge about system-level stuff. but since no one has reported this problem before, i don't suppose many people are crunching numbers on mac :)
anyways, this is my `gcc --version`:
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by blueberry** [comment]
No problem, that message was more intended to a future OSX user that need that issue solved faster. GCC version is only important to the one doing the actual compiling of the native library (currently *me*). After that, it should not matter what you have on the system if you only run the library.

**Comment by blueberry** [comment]
I upgraded to macOS Sierra, and upgraded Intel MKL to the brand new 2018.

The problem still persists for some single-precision float functions (on Mac only). The Neanderthal codebase is shared with Linux and Windows, so this is some Mac/MKL thing that need to be hunted down. All double-precision functions work as expected.

Here's the list of Neanderthal tests that fail. You'll note that it mostly affects dot, nrm2, asum, iamin, and some solver functions. 

```bash
FAIL "BLAS 1 dot." at (mkl_test.clj:90)
    Expected: 40.0
      Actual: 0.0

FAIL "BLAS 1 vector nrm2." at (mkl_test.clj:172)
Actual result did not agree with the checking function.
        Actual result: 1.0842022E-19
    Checking function: (roughly (Math/sqrt 14))

FAIL "BLAS 1 vector asum." at (mkl_test.clj:177)
    Expected: 6.0
      Actual: 0.0

FAIL "BLAS 1 GE asum." at (mkl_test.clj:376)
    Expected: 9.0
      Actual: 0.0

FAIL "BLAS 1 GE dot" at (mkl_test.clj:355)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a))

FAIL "BLAS 1 GE nrm2." at (mkl_test.clj:371)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly (Math/sqrt 19))

FAIL "BLAS 1 rot!" at (mkl_test.clj:132)
Actual result did not agree with the checking function.
        Actual result: 3.6893488147419103E19
    Checking function: (roughly 0 1.0E-6)

FAIL "BLAS 1 GE amax." at (mkl_test.clj:386)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 7.1)

FAIL "BLAS 1 Triangular dot" at (mkl_test.clj:641)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a))

FAIL "BLAS 1 Triangular dot" at (mkl_test.clj:642)
    Expected: 7.0
      Actual: 0.0

FAIL "BLAS 1 Triangular nrm2." at (mkl_test.clj:648)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (sqrt 55))

FAIL "BLAS 1 Triangular asum." at (mkl_test.clj:658)
    Expected: 9.0
      Actual: 0.0

FAIL "BLAS 1 Triangular amax." at (mkl_test.clj:653)
    Expected: 4.0
      Actual: 0.0

FAIL "LAPACK GE trf!" at (mkl_test.clj:1853)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0 0.02)

FAIL "LAPACK GE trs!" at (mkl_test.clj:2195)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0.0136)

FAIL "LAPACK GE trs!" at (mkl_test.clj:2196)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 0.012)

FAIL "LAPACK TR trs!" at (mkl_test.clj:2214)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0.0 1.0E-4)

FAIL "LAPACK GE trf/tri" at (mkl_test.clj:2226)
Actual result did not agree with the checking function.
        Actual result: -0.0
    Checking function: (roughly (sqrt 5.0))

FAIL "LAPACK TR tri" at (mkl_test.clj:2238)
Actual result did not agree with the checking function.
        Actual result: -1.0842022E-19
    Checking function: (roughly (sqrt 5.0))

FAIL "LAPACK TR tri" at (mkl_test.clj:2238)
Actual result did not agree with the checking function.
        Actual result: -1.0842022E-19
    Checking function: (roughly (sqrt 5.0))

FAIL "LAPACK GE con!" at (mkl_test.clj:2339)
Actual result did not agree with the checking function.
        Actual result: 0.03200535476207733
    Checking function: (roughly (/ 152.162))

FAIL "LAPACK TR con!" at (mkl_test.clj:2355)
Actual result did not agree with the checking function.
        Actual result: 0.031187040731310844
    Checking function: (roughly (con lu nrm))

FAIL "LAPACK TR con!" at (mkl_test.clj:2355)
Actual result did not agree with the checking function.
        Actual result: 0.031187040731310844
    Checking function: (roughly (con lu nrm))

FAIL "LAPACK GE sv!" at (mkl_test.clj:2266)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 0.01201)

FAIL "LAPACK GE sv!" at (mkl_test.clj:2267)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0.01359)

FAIL "LAPACK SY sv!" at (mkl_test.clj:2305)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0 1.0E-4)

FAIL "LAPACK SY linear system solver." at (mkl_test.clj:1908)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0 1.0E-4)

FAIL "LAPACK SY linear system solver." at (mkl_test.clj:1911)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 8)

FAIL "LAPACK SY linear system solver." at (mkl_test.clj:1914)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0 1.0E-4)

FAIL "LAPACK triangular banded linear system solver." at (mkl_test.clj:2090)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0.0 1.0E-4)

FAIL "LAPACK SY linear system solver." at (mkl_test.clj:1908)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0 1.0E-4)

FAIL "LAPACK SY linear system solver." at (mkl_test.clj:1911)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 8)

FAIL "LAPACK SY linear system solver." at (mkl_test.clj:1914)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0 1.0E-4)

FAIL "LAPACK GT linear system solver." at (mkl_test.clj:2136)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0 0.002)

FAIL "LAPACK GT linear system solver." at (mkl_test.clj:2137)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 0 0.002)

FAIL "LAPACK GT linear system solver." at (mkl_test.clj:2138)
Actual result did not agree with the checking function.
        Actual result: -3.6893488E19
    Checking function: (roughly 0 0.002)

FAIL "LAPACK GT linear system solver." at (mkl_test.clj:2136)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0 0.002)

FAIL "LAPACK GT linear system solver." at (mkl_test.clj:2137)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 0 0.002)

FAIL "LAPACK GT linear system solver." at (mkl_test.clj:2138)
Actual result did not agree with the checking function.
        Actual result: -3.6893488E19
    Checking function: (roughly 0 0.002)

FAIL "LAPACK GE qrf" at (mkl_test.clj:2389)
Actual result did not agree with the checking function.
        Actual result: -3.6893488E19
    Checking function: (roughly 0 0.015)

FAIL "LAPACK GE rqf" at (mkl_test.clj:2446)
Actual result did not agree with the checking function.
        Actual result: -3.6893488E19
    Checking function: (roughly 0 0.014)

FAIL "LAPACK GE lqf" at (mkl_test.clj:2481)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0 0.012)

FAIL "LAPACK GE qlf" at (mkl_test.clj:2514)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 0 0.014)

FAIL "LAPACK GE ls!" at (mkl_test.clj:2541)
Actual result did not agree with the checking function.
        Actual result: 1.0842022E-19
    Checking function: (roughly 0.003 1.0E-4)

FAIL "LAPACK GE ls!" at (mkl_test.clj:2543)
Actual result did not agree with the checking function.
        Actual result: -3.6893488E19
    Checking function: (roughly 0.0129)

FAIL "LAPACK GE lse!" at (mkl_test.clj:2564)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly 0.0 1.0E-4)

FAIL "LAPACK GE gls!" at (mkl_test.clj:2582)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0.0 1.0E-4)

FAIL "LAPACK GE ev!" at (mkl_test.clj:2615)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly 0.00943)

FAIL "LAPACK GE ev!" at (mkl_test.clj:2617)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0.01378)

FAIL "LAPACK GE ev!" at (mkl_test.clj:2618)
Actual result did not agree with the checking function.
        Actual result: -1.0842022E-19
    Checking function: (roughly 0.010403)

FAIL "LAPACK GE svd!" at (mkl_test.clj:2650)
Actual result did not agree with the checking function.
        Actual result: -3.6893488E19
    Checking function: (roughly 0.007526)

FAIL "LAPACK GE svd!" at (mkl_test.clj:2651)
Actual result did not agree with the checking function.
        Actual result: -2.0
    Checking function: (roughly 0.007526)

FAIL "LAPACK GE svd!" at (mkl_test.clj:2653)
Actual result did not agree with the checking function.
        Actual result: 3.6893488E19
    Checking function: (roughly 0.01586)

FAIL "LAPACK GE svd!" at (mkl_test.clj:2654)
Actual result did not agree with the checking function.
        Actual result: -1.0842022E-19
    Checking function: (roughly 0.012377)

FAIL "BLAS 1 SY dot" at (mkl_test.clj:888)
    Expected: 5.0
      Actual: 0.0

FAIL "BLAS 1 SY nrm2." at (mkl_test.clj:894)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (sqrt 81))

FAIL "BLAS 1 SY asum." at (mkl_test.clj:904)
    Expected: 13.0
      Actual: 0.0

FAIL "BLAS 1 SY amax." at (mkl_test.clj:899)
    Expected: 4.0
      Actual: 0.0

FAIL "BLAS 1 GB dot" at (mkl_test.clj:1112)
    Expected: -110.0
      Actual: 0.0

FAIL "BLAS 1 GB nrm2." at (mkl_test.clj:1118)
Actual result did not agree with the checking function.
        Actual result: 6.390139593977006E19
    Checking function: (roughly (sqrt 17))

FAIL "BLAS 1 GB asum." at (mkl_test.clj:1128)
    Expected: 153.0
      Actual: 0.0

FAIL "BLAS 1 SY dot" at (mkl_test.clj:888)
    Expected: 5.0
      Actual: 0.0

FAIL "BLAS 1 SY asum." at (mkl_test.clj:904)
    Expected: 13.0
      Actual: 0.0

FAIL "BLAS 1 Triangular dot" at (mkl_test.clj:641)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a))

FAIL "BLAS 1 Triangular dot" at (mkl_test.clj:642)
    Expected: 7.0
      Actual: 0.0

FAIL "BLAS 1 Triangular nrm2." at (mkl_test.clj:648)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (sqrt 55))

FAIL "BLAS 1 Triangular asum." at (mkl_test.clj:658)
    Expected: 9.0
      Actual: 0.0

FAIL "BLAS 1 Triangular amax." at (mkl_test.clj:653)
    Expected: 4.0
      Actual: 0.0

FAIL "BLAS 1 TP dot" at (mkl_test.clj:1325)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a))

FAIL "BLAS 1 TP dot" at (mkl_test.clj:1326)
    Expected: 7.0
      Actual: 0.0

FAIL "BLAS 1 Triangular asum." at (mkl_test.clj:658)
    Expected: 9.0
      Actual: 0.0

FAIL "BLAS 1 SP dot" at (mkl_test.clj:1357)
Actual result did not agree with the checking function.
        Actual result: NaN
    Checking function: (roughly (nrm2 a))

FAIL "BLAS 1 SP dot" at (mkl_test.clj:1358)
    Expected: 5.0
      Actual: -9.0

FAIL "BLAS 1 SY asum." at (mkl_test.clj:904)
    Expected: 13.0
      Actual: -5.0

FAIL "BLAS 1 GD nrm2." at (mkl_test.clj:1494)
    Expected: 2.236068
      Actual: -1.0842022E-19

FAIL "BLAS 1 GT dot" at (mkl_test.clj:1617)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a) 1.0E-6)

FAIL "BLAS 1 GT dot" at (mkl_test.clj:1618)
    Expected: -182.0
      Actual: 0.0

FAIL "BLAS 1 GT nrm2." at (mkl_test.clj:1625)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly (sqrt (dot a a)) 1.0E-5)

FAIL "BLAS 1 GT asum." at (mkl_test.clj:1635)
    Expected: 91.0
      Actual: 0.0

FAIL "BLAS 1 GT dot" at (mkl_test.clj:1617)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a) 1.0E-6)

FAIL "BLAS 1 GT dot" at (mkl_test.clj:1618)
    Expected: -182.0
      Actual: 0.0

FAIL "BLAS 1 GT nrm2." at (mkl_test.clj:1625)
Actual result did not agree with the checking function.
        Actual result: 2.0
    Checking function: (roughly (sqrt (dot a a)) 1.0E-5)

FAIL "BLAS 1 GT asum." at (mkl_test.clj:1635)
    Expected: 91.0
      Actual: 0.0

FAIL "BLAS 1 ST dot" at (mkl_test.clj:1714)
Actual result did not agree with the checking function.
        Actual result: 0.0
    Checking function: (roughly (nrm2 a) 1.0E-6)

FAIL "BLAS 1 ST dot" at (mkl_test.clj:1715)
    Expected: -310.0
      Actual: 0.0

FAIL "BLAS 1 GT nrm2." at (mkl_test.clj:1625)
Actual result did not agree with the checking function.
        Actual result: 3.6055512
    Checking function: (roughly (sqrt (dot a a)) 1.0E-5)

FAIL "BLAS 1 ST asum." at (mkl_test.clj:1726)
    Expected: 75.0
      Actual: 0.0
FAILURE: 87 checks failed.  (But 2381 succeeded.)
```

**Comment by MageMasher** [comment]
FYI this still exists with macOS Sierra 10.12.6 and m_mkl_2018.1.126.dmg Happy to help debug this issue if there is any way I can help.

**Comment by blueberry** [comment]
@MageMasher Yes there is. If possible, please experiment a bit with neanderthal-mkl (can be found inside neanderthal-native repository) and try to find why asum is 0.0 when a small array is the argument and some meaningless number when a large array is the input. 

I suspect two causes:
1. Either there is something wrong in how I compile it with clang on the macOS
2. or there is a bug in MKL

The easiest way is, of course, if you can set up a simple C project independent of neanderthal and try to compile the `asum` call so we can be sure.

**Comment by blueberry** [comment]
This was a bug in Intel MKL on macOS, which was fixed in MKL 2018.3. I haven't checked this, but since other projects detected the same issue, and it is solved in 2018.3, I'll close this without checking. If somehow this resurfaces, please feel free to reopen the issue.
